cpp/src/streaming/core/fanout.cpp

```diff
@@ -0,0 +1,430 @@
+        // it is guaranteed that message purging will be done only on indices less than
+        // next_idx, so we can safely send messages without locking the mtx
+        for (size_t i = state->ch_next_idx[idx]; i < curr_recv_msg_sz; i++) {
+            auto const& msg = state->recv_messages[i];
+            RAPIDSMPF_EXPECTS(!msg.empty(), "message cannot be empty");
+
+            // make reservations for each message so that it will fallback to host memory
+            // if needed
+            auto res = ctx.br()->reserve_or_fail(msg.copy_cost(), try_memory_types(msg));
+            if (!co_await ch_out->send(msg.copy(res))) {
+                // Failed to send message. Could be that the channel is shut down.
+                // So we need to abort the send task, and notify the process input task
+                co_await set_ch_idx_invalid.set_channel_idx_invalid();
+                co_return;
+            }
+        }
+        logger.trace(
+            "sent ", idx, " [", state->ch_next_idx[idx], ", ", curr_recv_msg_sz, ")"
+        );
+
+        // now next_idx can be updated to end_idx, and if !input_done, we need to request
+        // parent task for more data
+        auto lock = co_await state->mtx.scoped_lock();
+        state->ch_next_idx[idx] = curr_recv_msg_sz;
+        if (state->ch_next_idx[idx] == state->recv_messages.size()) {
+            if (state->input_done) {
+                // no more messages will be received, and all messages have been sent
+                break;
+            } else {
+                // request more data from the input channel
+                lock.unlock();
+                co_await state->request_data.notify_one();
+            }
+        }
+    }
+
+    co_await ch_out->drain(ctx.executor());
+    logger.trace("Send task ", idx, " completed");
+}
+
+/**
+ * @brief RAII helper class to set input_done to true and notify all send tasks to wind
+ * down when the unbounded fanout state goes out of scope.
+ */
+struct StateInputDoneAtExit {
+    std::shared_ptr<UnboundedFanoutState> state;
+
+    ~StateInputDoneAtExit() {
+        coro::sync_wait(set_input_done());
+    }
+
+    // forcibly set input_done to true and notify all send tasks to wind down
+    Node set_input_done() {
+        {
+            auto lock = co_await state->mtx.scoped_lock();
+            state->input_done = true;
+        }
+        co_await state->data_ready.notify_all();
+    }
+};
+
+/**
+ * @brief Wait for a data request from the send tasks.
+ *
+ * @param state The state of the unbounded fanout.
+ * @param last_completed_idx The index of the last completed message.
+ * @param latest_processed_idx The index of the latest processed message.
+ * @return True if the state is valid and can move forward, false otherwise (all send
+ * tasks are in an invalid state).
+ */
+auto wait_for_data_request(
+    UnboundedFanoutState& state, size_t& last_completed_idx, size_t& latest_processed_idx
```

Comment:
I would strongly suggest to return a std::tuple instead of the output parameters